[CSM-505] wallet shows incorrect balance again

pva701 opened the pull request:
This PR cancels calling `BListener` callbacks during LRC. Also, it slightly refactors code and removes some possible errors which don't yet appear.

pva701 commented on wallet/src/Pos/Wallet/Web/Tracking/Sync.hs:
```diff
@@ -387,52 +366,60 @@ trackingRollbackTxs (getEncInfo -> encInfo) allAddress getDiff getTs txs =
 
+type OwnTxInOuts = [((TxIn, TxOutAux), CWAddressMeta)]
+
+selectOwnTxInsAndOuts
```
this function performs common logic of `trackingApplyTxs` and `trackingRollbackTxs` and being just refactoring

pva701 commented on wallet/src/Pos/Wallet/Web/Tracking/Sync.hs:
```diff
@@ -387,52 +366,60 @@ trackingRollbackTxs (getEncInfo -> encInfo) allAddress getDiff getTs txs =
-            ownOutputAddrs = map cwamId ownOutputMetas
-
-            l = fromIntegral (length outs) :: Word32
-            ownTxIns = zip inps $ map fst ownInputs
```
this line was wrong because if there were some `Nothing` in tx undo (for unknown addresses) we would have concatenated some elements from `ownInputs` with wrong `TxIn`

kantp commented on node/src/Pos/Block/Slog/Logic.hs:
```diff
@@ -196,13 +197,17 @@ type MonadSlogApply ssc ctx m =
 
+-- | Flag determining whether to call BListener callback.
+newtype ShouldCallBListener = ShouldCallBListener Bool
+
```
Instead of wrapping a `Bool` inside a `newtype`, and then unwrapping it later to compare to `True` and `False`, I'd prefer something like `data ShouldCallBListener = CallBListener | DontCallBListener`, and then matching directly on those constructors.

Just a minor thing, though.

adinapoli-iohk replied:
I agree with Philip here, that would be even better 😉

pva701 commented:
Closed in favour #1777

gromakovsky commented on core/Pos/Aeson/Crypto.hs:
```diff
@@ -45,7 +55,39 @@ instance FromJSON PublicKey where
+    toJSON = toJSON . JsonByteString . Ed25519.unSignature
+
+instance FromJSON CC.XSignature where
+    parseJSON v = either (error . toText) identity . CC.xsignature . getJsonByteString <$> parseJSON v
```
Why `error`? Can you write a comment explaining why it will never happen?

gromakovsky commented on core/Pos/Aeson/Core.hs:
```diff
@@ -39,14 +49,56 @@ deriving instance ToJSON SlotCount
+deriveFromJSON defaultOptions ''SlotId
+deriveFromJSON defaultOptions ''BlockCount
+deriveFromJSON defaultOptions ''LocalSlotIndex
+deriveFromJSON defaultOptions ''ChainDifficulty
```
There are also `deriveToJSON`s for these types. I would merge them.

gromakovsky commented on core/Pos/Aeson/Core.hs:
```diff
@@ -39,14 +49,56 @@ deriving instance ToJSON SlotCount
+        "scrScript" .= JsonByteString scrScript ]
+
+instance FromJSON Script where
+    parseJSON = withObject "Script" $ \o -> do
```
I don't think it's a good idea to call a variable `o`, because in some editors/viewers it can look similar to `0`. Not important, just saying.

gromakovsky commented on core/Pos/Aeson/Crypto.hs:
```diff
@@ -45,7 +55,39 @@ instance FromJSON PublicKey where
 
 deriveJSON defaultOptions ''ProxySecretKey
+
+-- All these instances below needed for wallet
```
are needed

gromakovsky commented on core/Pos/Aeson/Crypto.hs:
```diff
@@ -45,7 +55,39 @@ instance FromJSON PublicKey where
+instance FromJSON CC.XSignature where
+    parseJSON v = either (error . toText) identity . CC.xsignature . getJsonByteString <$> parseJSON v
+
+instance ToJSON CC.XSignature where
```
Why are instances for such low-level types needed? 🤔

adinapoli-iohk replied:
Yea, was wondering the same actually.

gromakovsky commented on txp/Pos/Aeson/Txp.hs:
```diff
@@ -0,0 +1,99 @@
+        RedeemWitness{..} -> object
+            [ "tag" .= ("RedeemWitness" :: Text)
+            , "twRedeemKey" .= twRedeemKey
+            , "twRedeemSig" .= twRedeemSig
```
Tbh, I don't get why you decided to prepend `tw` prefixes, they look like unnecessary extra verbosity to me :/

adinapoli-iohk replied:
I concur!

gromakovsky commented on wallet/src/Pos/Aeson/Storage.hs:
```diff
@@ -0,0 +1,64 @@
+-- | These instances are used for `test/printDB` method.
+-- Only for debugging and testing purposes.
+
+module Pos.Aeson.Storage where
```
Where is (empty) export list? :kot:

adinapoli-iohk commented on auxx/src/Command/Rollback.hs:
```diff
@@ -55,7 +58,7 @@ rollbackAndDump numToRollback outFile = do
             logInfo $ sformat ("Dumped "%int%" transactions to "%string)
                       (length allTxs) (outFile)
-            rollbackBlocksUnsafe (BypassSecurityCheck True) blunds
+            rollbackBlocksUnsafe (BypassSecurityCheck True) (ShouldCallBListener True) blunds
```
EDIT: Philipp's comment was preserved (sorry, I thought it got lost). Ignore this comment then 😉 

A minor one but I still think @kantp 's idea was nice enough to be ported-over here: Why not coalesce that `Bool` into a sum type directly, like the following?

```
data BListenerAction = Call
                                     | DontCall
```

(I suck at naming things, but you get the idea). Essentially, wrapping a `Bool` in an another type constructor will give you only marginal gains in terms of type safety (although _will_ cure boolean blindness). At this point, why not scrap the wrapped `Bool` entirely?

Nothing that will stop approving the PR at this stage, but I'm curious to hear what people think!

gromakovsky replied:
I like `data ShouldCallBListener = CallBListener | DontCallBListener`, but please don't do it in this PR, because in master we already have `ShouldCallBListener Bool`.

adinapoli-iohk replied:
Fair enough, in which case I would be happy enough with a new issue being created on YT as commitment we are going to do that 😀

adinapoli-iohk commented on core/Pos/Aeson/Core.hs:
```diff
@@ -39,14 +49,56 @@ deriving instance ToJSON SlotCount
 instance FromJSON CoinPortion where
     parseJSON v = unsafeCoinPortionFromDouble <$> parseJSON v
 
+instance ToJSON CoinPortion where
```
I have yet to arrive at the bottom of the PR, so I beg your pardon if this is contemplated later, but essentially I would like to see, every time we add _separated_ `ToJSON` and `FromJSON` instances (i.e. not generated atomically like in `deriveJSON`), roundtrip properties (as in `decode $ encode myCoinPortion === Just myCoinPortion`). As we are fiddling with doubles (it seems) I think it's important to have confidence we are not losing precision. Similar considerations can be applied to `Timestamp`.

adinapoli-iohk commented on core/Pos/Aeson/Core.hs:
```diff
@@ -39,14 +49,56 @@ deriving instance ToJSON SlotCount
+instance ToJSON Script where
+    toJSON Script{..} = object [
+        "scrVersion"    .= scrVersion,
+        "scrScript" .= JsonByteString scrScript ]
```
I tend to agree with @gromakovsky 's comment below: why prefix the JSON fields with `scr` (or any other prefix)? . I think `version` and `script` looks way neater.

adinapoli-iohk commented on core/Pos/Util/Modifier.hs:
```diff
@@ -161,6 +163,27 @@ delete
+    -> MapModifier k v
+alter f key (MapModifier mm) = MapModifier $ HM.alter transformedF key mm
+  where
+    kiToMaybe KeyNotFound     = Nothing
```
As this is a local binding, why not simply name it `toMaybe`?

adinapoli-iohk commented on core/Pos/Util/Modifier.hs:
```diff
@@ -161,6 +163,27 @@ delete
+    -> k
+    -> MapModifier k v
+    -> MapModifier k v
+alter f key (MapModifier mm) = MapModifier $ HM.alter transformedF key mm
```
I would suggest giving `alter` the same type signature of `HM.alter`, for the _principle of least surprise_, as `alter` is defined (for `HashMap`) as:

```
(Eq k, Hashable k) => (Maybe v -> Maybe v) -> k -> HashMap k v -> HashMap k v
```

I think will also simplify your implementation, which is now required to do all this double wrapping.

pva701 replied:
tbh I don't like types like `Maybe (Maybe v)`, it's not clear what it means. also `KeyState` describes a state of key very clearly.

adinapoli-iohk commented on wallet/src/Pos/Wallet/Web/Api.hs:
```diff
@@ -386,6 +394,8 @@ type WalletApi = ApiPrefix :> (
      -- NOTE: enabled in prod mode https://issues.serokell.io/issue/CSM-333
      TestReset
     :<|>
+     TestState
```
Surely there is some mechanism which does this automatically, but how can we prevent that this endpoint will leak to production as it's for debugging-purposes only?

pva701 replied:
we can add flag `--with-test-endpoints` which will include these `/test` endpoints and turn off it in production

adinapoli-iohk commented on wallet/src/Pos/Wallet/Web/Swagger/Description.hs:
```diff
@@ -52,6 +52,9 @@ instance HasCustomSwagger TestReset where
         "Delete all secret keys. It works in development mode only, \
         \returns HTTP 403 otherwise."
 
+instance HasCustomSwagger TestState where
```
This is a philosophical observation, and actually food for though for the upcoming wallet re-design: It feels wrong exposing internal/tests endpoints to the swagger documentation, and it's confusing for 3rd parties to see. We should probably find out a way to still document this (for backend developers) without leaking it to end-users docs.

For now, I think that's fine.

adinapoli-iohk commented on wallet/src/Pos/Wallet/Web/Tracking/Modifier.hs:
```diff
@@ -103,6 +103,15 @@ type CachedCAccModifier = CAccModifier
 -- Funcs
 ----------------------------------------------------------------------------
 
+deleteNotDeep :: (Eq k, Hashable k) => k -> MapModifier k v -> MapModifier k v
```
This function would benefit from a top-level comment to explain what "deep" means in the context.

adinapoli-iohk commented on wallet/src/Pos/Wallet/Web/Tracking/Sync.hs:
```diff
@@ -378,7 +357,7 @@ trackingApplyTxs (getEncInfo -> encInfo) allAddresses getDiff getTs getPtxBlkInf
     :: forall ssc . (HasConfiguration, SscHelpersClass ssc)
     => EncryptedSecretKey -- ^ Wallet's secret key
-    -> [CWAddressMeta] -- ^ All adresses
+    -> [CWAddressMeta]    -- ^ All adresses
```
Typo, it should be `addresses`.

dcoutts commented:
This is now included in #1815.

gromakovsky commented:
It is not _fully_ included in #1815 (e. g. there is no `Pos.Aeson.Txp` module in that PR) and also some comments haven't been addressed, so I am reopening.

pva701 commented:
I've addressed all issues. There are two problems: absence of roundtrip tests and visibility of test endpoints. JSON instances are used for `test/state` which is debug endpoint, so I propose to postpone writing roundtrip tests and do it as separate issues.

@gromakovsky @adinapoli-iohk @dcoutts re-review plz
I've heard we planned to publish an update on Monday, I would like these changes to get into this update.

adinapoli-iohk commented on core/Pos/Aeson/Core.hs:
```diff
@@ -63,21 +63,15 @@ deriveToJSON   S.defaultOptions ''Microsecond
     toJSON Script{..} = object [
-        "scrVersion"    .= scrVersion,
-        "scrScript" .= JsonByteString scrScript ]
+        "rersion"    .= scrVersion,
```
Is this correct? Surely it should be `version`? 😀

gromakovsky replied:
Typo.
